### PR TITLE
Modernize sphinx

### LINKED ISSUE #673: Deprecations don't properly format properties
Deprecation of [`faceGradAverage`](https://www.ctcms.nist.gov/fipy/fipy/generated/fipy.variables.html#fipy.variables.cellVariable.CellVariable.faceGradAverage) should be `grad.arithmeticFaceValue`, not `grad.arithmeticFaceValue()`

### LINKED ISSUE #865: Sphinx search is broken on website
https://www.ctcms.nist.gov/fipy/search.html doesn't return anything

### PR DESCRIPTION
**Pull request by guyer**
Clean up legacy Sphinx hacks and resolve longstanding warnings.

Support use of [NISTtheDocs2Death](https://github.com/usnistgov/NISTtheDocs2Death) to put documentation on https://pages.nist.gov/fipy.

Previous configuration went to great lengths to "see" things outside documentation directory, which is not the intended design of sphinx. Using `autodoc` instead of `autosummary` simplifies things considerably. 

- [x] Search box does not appear (Fixes #865)
- [x] Fixes #673

Test failures are only nix.

> [!NOTE]
> The remaining warning is because mayavi isn't installed.
>
> Attempting to use `docs/environment.yml` causes a 30 min delay.
>
> Adding mayavi to `docs/requirements.txt` slows down installation by 10 min and doesn't resolve the error.
> 
> (petsc was fixed with `autodoc_mock_imports`, but attempting to add petsc4py to `docs/requirements.txt`
> fails when trying to build wheel.)

### CODE REVIEW

**Comment by guyer** [comment]
See https://pages.nist.gov/fipy/en/modernize_sphinx for build results

**Comment by guyer** [comment]
> [!IMPORTANT]  
> Search box does not appear

File: examples/benchmarking/benchmarker.py
Comment (wd15) [inline examples/benchmarking/benchmarker.py:1]: Probably should just delete this

File: fipy/tools/numerix.py [hunk 1]
```diff
@@ -222,10 +226,11 @@ def rank(a):
 
     .. note::
 
-       The rank of a `MeshVariable` is for any single element. E.g., A
-       `CellVariable` containing scalars at each cell, and defined on a 9
-       element `Grid1D`, has rank 0. If it is defined on a 3x3 `Grid2D`, it is
-       still rank 0.
+       The rank of a :class:`~fipy.variables.meshVariable.MeshVariable` is
```
Comment (wd15) [inline fipy/tools/numerix.py:229]: Why am I seeing [this](https://pages.nist.gov/fipy/en/modernize_sphinx/generated/fipy.tools.numerix.html#fipy.tools.numerix.rank) on pages.nist.gov?

Comment (wd15) [reply fipy/tools/numerix.py:229]: Sorry, all good. I'm seeing the correct version

**Comment by wd15** [comment]
I can't find any issues when randomly checking the changes and the rendered version on pages.nist.gov. I'll try and fix the nix tests at some point.

**Review by wd15 — APPROVED** [review]

**Comment by guyer** [comment]
> I can't find any issues when randomly checking the changes and the rendered version on pages.nist.gov. I'll try and fix the nix tests at some point.

Much obliged @wd15

File: fipy/viewers/mayaviViewer/mayaviDaemon.py [hunk 2]
```diff
@@ -316,8 +316,10 @@ def view_data(self):
 signal.signal(signal.SIGTERM, MayaviDaemon._sigint_handler)
 
 def main(argv=None):
-    """Simple helper to start up the mayavi application.  This returns
-    the running application."""
+    """Simple helper to start up the mayavi application.=
```
Comment (tkphd) [inline fipy/viewers/mayaviViewer/mayaviDaemon.py:319]:
```suggestion
    """Simple helper to start up the mayavi application.
```

**Review by tkphd — APPROVED** [review]
Ah-_Ha!_ I found the typo!

**Comment by guyer** [comment]
> Ah-_Ha!_ I found the typo!

Curses!

**Comment by tkphd** [comment]
The PDF docs build, but there are a lot of empty "chapters" for Examples:



The links in the chapter TOCs take the reader to the `fipy Package Documentation` examples section, some 800 pages further into the manual.

**Comment by guyer** [comment]
> The PDF docs build, but there are a lot of empty "chapters" for Examples:
> 
> The links in the chapter TOCs take the reader to the `fipy Package Documentation` examples section, some 800 pages further into the manual.

Well, they're not empty; they're just lists of links. It's not feasible to render the examples twice, nor am I sure it would be a good idea to do so. 

The point of the "**II. Examples**" part is to give a curated list of examples that will help teach users how to use FiPy. The point of "**III. `fipy` Package Documentation > Examples**" is to provide a complete API listing of all of the examples that are used to test the higher functionality of FiPy.

The way we used to do it, with just the curated examples and no API listing of all examples generated lots of errors and warnings when I started using sphinx "properly".

Is there a better way to present this? If I incorporated "**II. Examples**" as a chapter in "**I. Introduction**", then the sections would still be "empty", but they wouldn't be chapters. Would that be better?

Frankly, I'm starting to wonder if a 1200 page PDF isn't just useless and stupid.

**Comment by tkphd** [comment]
Fair. Perhaps include verbiage in II reflecting the fact that these are lists of links, deliberately devoid of content?

**Comment by guyer** [comment]
PETSc breakages [linux-py3k-petsc](https://dev.azure.com/guyer/FiPy/_build/results?buildId=844&view=logs&j=99bad03a-9a60-5e68-7669-1870cb1c1434), [linux-py3k-petsc-parallel](https://dev.azure.com/guyer/FiPy/_build/results?buildId=844&view=logs&j=58fde56f-e4bb-53ea-f269-cde8ff5cac60), [macos-py3k-petsc](https://dev.azure.com/guyer/FiPy/_build/results?buildId=844&view=logs&j=d140db5a-ea74-5114-89c8-047f038a6232&t=d140db5a-ea74-5114-89c8-047f038a6232), [macos-py3k-petsc-parallel](https://dev.azure.com/guyer/FiPy/_build/results?buildId=844&view=logs&j=6e9fd3b2-3876-5f92-ebcc-0273b412dd37) unrelated to this PR. They correspond to [PETSc 3.20.0](https://dev.azure.com/guyer/FiPy/_build/results?buildId=844&view=logs&j=6e9fd3b2-3876-5f92-ebcc-0273b412dd37&t=9e30d64e-08f3-56bc-e828-0c4d9a37260a&l=220), (released [2023-09-28](https://petsc.org/release/changes/), conda-forge [petsc4py feedstock updated 2023-10-04](https://github.com/conda-forge/petsc4py-feedstock/commits/main)). Previous [PETSc 3.18.4](https://dev.azure.com/guyer/FiPy/_build/results?buildId=838&view=logs&j=d140db5a-ea74-5114-89c8-047f038a6232&t=e01c99fe-a45f-5995-24fb-d520d2bba4ff&l=213) worked fine.

**Pull request merged**